// @shelf/jest-mongodb/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMongoMemoryReplSetOptions = isMongoMemoryReplSetOptions;
exports.getMongodbMemoryOptions = getMongodbMemoryOptions;
exports.getMongoURLEnvName = getMongoURLEnvName;
exports.shouldUseSharedDBForAllJestWorkers = shouldUseSharedDBForAllJestWorkers;
const child_process_1 = require("child_process");
const path_1 = require("path");
function isMongoMemoryReplSetOptions(options) {
    return Boolean(options?.replSet);
}
function getConfigFile() {
    return process.env.MONGO_MEMORY_SERVER_FILE || 'jest-mongodb-config.js';
}
const configCache = new Map();
function importConfig(configPath) {
    try {
        const { status, stdout } = (0, child_process_1.spawnSync)(process.execPath, [
            '--input-type=module',
            '--eval',
            [
                'import {pathToFileURL} from "node:url";',
                `const mod = await import(pathToFileURL(${JSON.stringify(configPath)}).href);`,
                'const payload = mod.default ?? mod;',
                'console.log(JSON.stringify(payload));',
            ].join('\n'),
        ], { encoding: 'utf8' });
        if (status === 0 && stdout.trim()) {
            return JSON.parse(stdout);
        }
    }
    catch {
        // ignore and fall through to undefined
    }
    return undefined;
}
function loadConfig(cwd) {
    const baseDir = cwd || process.cwd();
    if (configCache.has(baseDir)) {
        return configCache.get(baseDir) ?? undefined;
    }
    const configPath = (0, path_1.resolve)(baseDir, getConfigFile());
    try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const loadedConfig = require(configPath);
        if (loadedConfig && typeof loadedConfig.default !== 'undefined') {
            const config = loadedConfig.default;
            configCache.set(baseDir, config ?? null);
            return config;
        }
        const config = loadedConfig;
        configCache.set(baseDir, config ?? null);
        return config;
    }
    catch {
        const importedConfig = importConfig(configPath);
        configCache.set(baseDir, importedConfig ?? null);
        return importedConfig;
    }
}
function getMongodbMemoryOptions(cwd) {
    const config = loadConfig(cwd);
    if (config?.mongodbMemoryServerOptions) {
        return config.mongodbMemoryServerOptions;
    }
    return {
        binary: {
            checkMD5: false,
        },
        instance: {},
    };
}
function getMongoURLEnvName(cwd) {
    const config = loadConfig(cwd);
    return config?.mongoURLEnvName || 'MONGO_URL';
}
function shouldUseSharedDBForAllJestWorkers(cwd) {
    const config = loadConfig(cwd);
    if (typeof config?.useSharedDBForAllJestWorkers === 'undefined') {
        return true;
    }
    return config.useSharedDBForAllJestWorkers;
}
